#Remove all objects (i.e., data sets) from R
rm(list=ls())

###Required files:
### (1) imputedMarkers.allchr.0.1cm.final.Panzea.consolidated.B.txt (GBS SNPs)
### (2) GAPIT functions
### (3) GWAS result summary ("Complete Results...")generated by script "Merge_Chr_Data_Sets_into_One_File_HMv1v2_with_window_GITHUB.r"


#Read in the gneotypic data
#setwd("F:\\Permutations\\GBS_SNPs\\imputedMarkers.allchr.0.1cm.final.Panzea.consolidated.B.txt")
#genotypes <- read.table("imputedMarkers.allchr.0.1cm.final.Panzea.consolidated.B.txt", head = TRUE)

genotypes <- read.table("C:\\Users\\ceb19\\Documents\\Gore Lab\\Carotenoid NAM Merged Env\\(9)JL Analysis\\Permutations\\GBS_SNPs\\imputedMarkers.allchr.0.1cm.final.Panzea.consolidated.B.txt", head = TRUE)
JL.path <- "C:\\Users\\ceb19\\Documents\\Gore Lab\\Carotenoid NAM Merged Env\\(9)JL Analysis\\Permutations\\Data_for_1pct_Support_Intervals\\JL_output\\JL_model_modified_SI01\\"

#Step 0: Set directory to load GAPIT source files (this step is omited for using package)
#######################################################################################
library('MASS')
library(multtest)
library(gplots)
library("compiler")

#Set your working directory
#Set your working directory

#setwd("F:\\NAM_Carot_GWAS_v1v2")

#working.directory <- "F:\\NAM_Carot_GWAS_v1v2\\"
source("http://www.maizegenetics.net/images/stories/bioinformatics/GAPIT/emma.txt")

source("http://www.maizegenetics.net/images/stories/bioinformatics/GAPIT/gapit_functions.txt")
#source("test.r")
#load("GAPIT.EMMA.Workspace20120120_FOR_EMMAX_P3D.Rdata")
#source("C:\\Users\\ael54\\Desktop\\August_2010_data\\GAPIT\\Add_R_square_and_BIC\\Modified_Code_from_the_Internet.r")

#source("EMMAxP3D_20111101 - Copy.r")

##############Note: Adapt this loop so that mutliple traits can be run to make these kind of plots:

#Input a list of paths, and use the "List" function to combine them into one object

trait<- "BCAR_RUV" 

#trait<- c("LUT_RUV","ZEA_RUV","ZEI_RUV","BCRY_RUV",
#"ACAR_RUV","BCAR_RUV","PHYF_RUV","THLYC_RUV",
#"LUT_MSS","ZEA_MSS","ZEI_MSS","BCRY_MSS",
#"ACAR_MSS","BCAR_MSS","PHYF_MSS","THLYC_MSS",
#"TOTCAR_RUV","TOTCAR_MSS") 


#######################
#Note that MSS files and 
#both totcar files did 
#not produce plots
########################
          
trait.multicollinearity <- c("BCAR_RUV")


# c("LUT_RUV","ZEA_RUV","ZEI_RUV","BCRY_RUV",
#"ACAR_RUV","BCAR_RUV","PHYF_RUV","THLYC_RUV",
#"LUT_MSS","ZEA_MSS","ZEI_MSS","BCRY_MSS",
#"ACAR_MSS","BCAR_MSS","PHYF_MSS","THLYC_MSS",
#"TOTCAR_RUV","TOTCAR_MSS")
#


max.log.pval <- 40 #Maximum -log10(P-value) to include on the graphs.

 #Create Marker cateogries. I will comment these out for now, but we may want to put this back in later.
#hm2.missing <- c("A/C_as_missing", "A/T_as_missing", "A/G_as_missing", 
#             "C/A_as_missing", "C/T_as_missing", "C/G_as_missing", "T/A_as_missing", "T/C_as_missing", "T/G_as_missing", 
#             "G/A_as_missing", "G/C_as_missing", "G/T_as_missing")

#hm2.minor <- c("A/C_as_minor", "A/T_as_minor", "A/G_as_minor", "C/A_as_minor",
#           "C/T_as_minor", "C/G_as_minor", "T/A_as_minor", "T/C_as_minor", "T/G_as_minor", 
#           "G/A_as_minor", "G/C_as_minor", "G/T_as_minor")
#
#hm1 <- c("A/C_hm1", "A/T_hm1", "A/G_hm1", "C/A_hm1", "C/T_hm1", "C/G_hm1", "T/A_hm1", "T/C_hm1", "T/G_hm1", "G/A_hm1", "G/C_hm1", "G/T_hm1")
#
#cnv <- "cnv-" 

jitter.JL <- 100000000


#######################################
### NOTE: Script has been modified by AEL on Nov 6 2013
### some associations lie close to the line demarcating the boundary of 2 chromosomes
### in addition, the "end" of the chromosome in these graphs is defined by the last QTL detected,
### NOT the true, physical end of the chromosome
### as a result, the data are obscured
### Jitter of 100 Mb was added to the last QTL so that the data could be visualized
### Plan is to add a map file of the maize containing physical coordinates
########################################

#For each path (i.e., for each trait)


for(i in trait){

##################



#Phenotypic Data
#data  <- read.table(paste(working.directory,i,"\\Complete_Results_",i,".txt", sep = ""), head = TRUE)
### NOTE that code below has been changed to be specific to BCAR alone and will need to be modified for entire loop of traits
data  <- read.table(paste("C:\\Users\\ceb19\\Documents\\Gore Lab\\Carotenoid NAM Merged Env\\(10)GWAS Analysis\\RUV GWAS 25fam_HMPonly - rerun\\BCAR_RUV_no_collin_25fam_HMPonly\\Complete_Results_BCAR_RUV_no_collin_25fam_HMPonly.txt", sep = ""), head = TRUE)

NA.Fval <- rep(NA, nrow(data))

data <- cbind(data, NA.Fval)
colnames(data)[5] <- "Fval"

##if(i %in% trait.multicollinearity){
  #Read in JL Results from R (where the model with no multicollinearity was fitted)
  #JL.results <- read.table(paste(JL.path,"Investigating_Multicollinearity\\", i,"_JL_model_modified_no_multicollinearity_SI01.txt", 
                                 #sep = ""), head = TRUE)
  ##JL.results <- read.table(paste(JL.path, i,"_JL_model_modified_no_multicollinearity_SI01.txt", 
                              ##   sep = ""), head = TRUE)
  
  #Remove the first and last rows, as these are not marker estimates
 ## JL.results <- JL.results[-which((JL.results[,1] == "pop")|(JL.results[,1] == "Residuals")),]
  
  #Obtain the SNP ids
 ## SNP.ids <- substr(as.character(JL.results[,1]), 5, nchar(as.character(JL.results[,1])))
  
  #Obtain the chromosome number from the genotype file
 ## chr <- NULL
 ## for(j in 1:length(SNP.ids)){
  ##  chr <- c(chr, genotypes[which(as.character(genotypes[,1]) == SNP.ids[j]),3])
 ## }
  
  #Obtain the bp position from the genotype file
 ## pos <- NULL
 ## for(j in 1:length(SNP.ids)){
  ##  pos <- c(pos, genotypes[which(as.character(genotypes[,1]) == SNP.ids[j]),4])
 ## }
 
  #Obtain the -log10 P-values
 ## log.pf <- -log10(JL.results[,6])
  
##}else{

##################


  #Read in JL Results
  JL.results <- read.table(paste(JL.path,i,"_JL_model_modified_no_multicollinearity_SI01.txt", 
                                 sep = ""), head = TRUE)
                                    
  #Truncate the JL results so that you have only what you need
  JL.results.truncated <- JL.results[-1,c(2,4,9)]


  #Reformat to extract only the information that you need.
  chr <- JL.results.truncated[,1]
  pos <- as.numeric(as.vector( substr(JL.results.truncated[,2],3,nchar(as.character(JL.results.truncated[,2])))))
  log.pf <- -log10(JL.results.truncated[,3])
##}

if(length(which(log.pf > max.log.pval)) > 0)   log.pf[which(log.pf > max.log.pval)] = max.log.pval

JL.for.data <- cbind(chr,pos,rep(NA,length(chr)),rep(NA,length(chr)),log.pf)
colnames(JL.for.data) <- colnames(data) 


data <- rbind(data, JL.for.data)

#Sort the phenotypic data by chromosome and bp position (to do this, we will sort by bp and then by chromosome)
data <- data[order(data[,2]),]
data <- data[order(data[,1]),]

######################################################################################################
# Manipulate the following code to make the plots that Dean requested.

#Create a color vector
#Similar to lines  36-47, I am temporarily commenting out these lines.
#color.vector <- rep("black", nrow(data))
#color.vector[which(data[,3]%in%hm2.missing)] = "lightblue"
#color.vector[which(data[,3]%in%hm2.minor)] = "blue"
#color.vector[which(data[,3]%in%hm1)] = "grey"
#color.vector[which(data[,3]%in%cnv)] = "green"

chm.to.analyze <- c(1,2,3,4,5,6,7,8,9,10)
lastbase <- 0
ticks <- NULL

  #change base position to accumulatives
  for (j in chm.to.analyze)
  {
    index=(data[,1]==j)
    ticks <- c(ticks, lastbase+mean(data[index,2]))
    data[index,2]=data[index,2]+lastbase
    lastbase=max(data[index,2])
  }


ticks.end <- NULL
for(j in chm.to.analyze[-length(chm.to.analyze)]){

  data.subset <- data[which(data[,1]==j),]
  ticks.end <- c(ticks.end, max(data.subset[,2]))
  
  
  #If the JL peak is the first locus, move the bp position forward by 500kb. The second logical statement is there
    #"(nrow(data.subset)!=1)" for the situation when there is only one JL peak from a particular chromosome
  if((!is.na(data.subset[1,4]))&(nrow(data.subset)!=1)){
   data[which(data[,1]==j)[1],2] =  data[which(data[,1]==j)[1],2]+jitter.JL
  }
    
  #If the JL peak is the last locus, move the bp position backward by 500kb
  if(!is.na(data.subset[nrow(data.subset),4])){
   data[which(data[,1]==j)[length(which(data[,1]==j))],2] =  data[which(data[,1]==j)[length(which(data[,1]==j))],2]-jitter.JL
  }
    
}


  #Set the limits of the y (RMIP) and z (-log10(P-value)) coordinates

  y.lim <- ceiling(max(data[which(!is.na(data[,4])),4]))/100 

  z.lim <- ceiling(max(data[which(!is.na(data[,5])),5]))


   x <- data[ , 2]
   y <- data[ , 4]/100
   z <- data[ , 5]

   ticks.z <- c(0, max.log.pval/4, max.log.pval/2,(3*max.log.pval)/4, max.log.pval)
   labels.z <- c(0, max.log.pval/4, max.log.pval/2,(3*max.log.pval)/4, "max")
   
#print("Manhattan XY created")

  #pdf(paste(working.directory,i,"\\JL_GWAS_Summary_Plot_",i,".pdf", sep = ""), width = 10)
  pdf(paste("C:\\Users\\ceb19\\Documents\\Gore Lab\\Carotenoid NAM Merged Env\\(10)GWAS Analysis\\RUV GWAS 25fam_HMPonly - rerun\\BCAR_RUV_no_collin_25fam_HMPonly\\JL_GWAS_Summary_Plot_",i,".pdf", sep = ""), width = 10)
  #pdf(paste("Desktop\\JL_GWAS_Summary_Plot_",i,".pdf", sep = ""), width = 10)
  par(mar = c(5,5,5,5))

   plot(x,y, pch = 17, col = "purple", ylim = c(0,1), axes = FALSE, xlab = "", ylab = "")
   axis(1, at=ticks,cex.axis=1.5,labels=chm.to.analyze,tick=F)
   mtext("Chromosome", side=1,line=3, cex = 1.5)
   axis(2, at=c(0,0.25, 0.5, 0.75, 1),cex.axis=1.5,labels=c("0.0","0.25","0.5","0.75","1.0"),tick=F)
   mtext("GWAS: Marker Significance", side=2,line=3, cex = 1.5)
   par(new = TRUE)
   plot(x,z, type="h",col="darkgreen",xaxt="n",yaxt="n",xlab="",ylab="", lwd = 2, ylim = c(0,z.lim))
   axis(4, at = ticks.z, labels=labels.z, cex.axis = 1.5)
   mtext("JL: -log(P-value)", side=4,line=3, cex = 1.5)
   abline(v = ticks.end, col = "lightgrey", lwd = 1.5, cex = 1.0)
   legend("topleft", c("JL QTL Peak", "GWAS SNP"), col = c("darkgreen","purple"), pch = c(20, 17), cex=1.5,  pt.cex = 3, bty = "n")
   box()

dev.off()


rm(data)

  ###END MODIFICATION


#############################Now, make a graph for the sum(RMIP) results within each window

#Phenotypic Data
data  <- read.table(paste(working.directory,i,"\\Complete_Results_within_ME_Windows_",i,".txt", sep = ""), head = TRUE)

NA.Fval <- rep(NA, nrow(data))

data <- cbind(data, NA.Fval)
colnames(data)[4] <- "Fval"


JL.for.data <- cbind(chr,pos,rep(NA,length(chr)),log.pf)
colnames(JL.for.data) <- colnames(data) 


data <- rbind(data, JL.for.data)


#Sort the phenotypic data by chromosome and bp position (to do this, we will sort by bp and then by chromosome)
data <- data[order(data[,2]),]
data <- data[order(data[,1]),]

######################################################################################################
# Manipulate the following code to make the plots that Dean requested.

#Create a color vector
#Similar to lines  36-47, I am temporarily commenting out these lines.
#color.vector <- rep("black", nrow(data))
#color.vector[which(data[,3]%in%hm2.missing)] = "lightblue"
#color.vector[which(data[,3]%in%hm2.minor)] = "blue"
#color.vector[which(data[,3]%in%hm1)] = "grey"
#color.vector[which(data[,3]%in%cnv)] = "green"

chm.to.analyze <- c(1,2,3,4,5,6,7,8,9,10)
lastbase <- 0
ticks <- NULL

  #change base position to accumulatives
  for (j in chm.to.analyze)
  {
    index=(data[,1]==j)
    ticks <- c(ticks, lastbase+mean(data[index,2]))
    data[index,2]=data[index,2]+lastbase
    lastbase=max(data[index,2])
  }


ticks.end <- NULL
for(j in chm.to.analyze[-length(chm.to.analyze)]){
  data.subset <- data[which(data[,1]==j),]
  ticks.end <- c(ticks.end, max(data.subset[,2]))
  
#If the JL peak is the first locus, move the bp position forward by 500kb. The second logical statement is there
    #"(nrow(data.subset)!=1)" for the situation when there is only one JL peak from a particular chromosome
  if((!is.na(data.subset[1,4]))&(nrow(data.subset)!=1)){
   data[which(data[,1]==j)[1],2] =  data[which(data[,1]==j)[1],2]+jitter.JL
  }
    
  #If the JL peak is the last locus, move the bp position backward by 500kb
  if(!is.na(data.subset[nrow(data.subset),4])){
   data[which(data[,1]==j)[length(which(data[,1]==j))],2] =  data[which(data[,1]==j)[length(which(data[,1]==j))],2]-jitter.JL
  }
  
  
  
}


  #Set the limits of the y (RMIP) and z (-log10(P-value)) coordinates

  y.lim <- ceiling(max(data[which(!is.na(data[,3])),3]))/100 

  #z.lim <- ceiling(max(data[which(!is.na(data[,5])),5]))

   x <- data[ , 2]
   y <- data[ , 3]
   z <- data[ , 4]


   window.size.genetic <- substr(colnames(data)[2], 13, (nchar(colnames(data)[2])-7))
   
    
#print("Manhattan XY created")

  pdf(paste(working.directory,i,"\\JL_GWAS_Summary_Plot_within_",window.size.genetic,"Windows",i,".pdf", sep = ""), width = 10)
  par(mar = c(5,5,5,5))

   plot(x,y, pch = 19, col = "darkgreen", axes = FALSE, xlab = "", ylab = "", main = paste("For GWAS: Genome Subdivided into ",window.size.genetic, " Bins" , sep = ""), cex.main = 1.5)
   axis(1, at=ticks,cex.axis=1.5,labels=chm.to.analyze,tick=F)
   mtext("Chromosome", side=1,line=3, cex = 1.5)
   axis(2,cex.axis=1.5,tick=F)
   mtext("GWAS: Sum of Results within Windows", side=2,line=3, cex = 1.5)
   par(new = TRUE)
   plot(x,z, type="h",col="purple",xaxt="n",yaxt="n",xlab="",ylab="", lwd = 2, ylim = c(0,z.lim))
   axis(4, at = ticks.z, labels=labels.z, cex.axis = 1.5)
   mtext("JL: -log(P-value)", side=4,line=3, cex = 1.5)
   abline(v = ticks.end, col = "lightgrey", lwd = 1.5, cex = 1.0)
   legend("topleft", c("JL QTL Peak", "GWAS Bin Midpoint"), col = c("purple", "darkgreen"), pch = c(20, 20), cex=1.5, bty = "n", pt.cex = 3)
   box()

   
   
   
   #Delete the above code


dev.off() 
 

}#End for(i in 1:length(trait.name))

#
